File: 566. Reshape the Matrix/566.go
package  main

import "fmt"

func main(){
	fmt.Println(matrixReshape([][]int{{1,2},{3,4}},2,4))
}

func matrixReshape(nums [][]int, r int, c int) [][]int {
    if len(nums)*len(nums[0]) != r*c{
		newnums := make([][]int, len(nums))
		for i := 0; i < len(nums); i++ {
			newnums[i] = make([]int, len(nums[0]))
			copy(newnums[i], nums[i])
		}
		return newnums
	}

	onenums := make([]int, r*c)
	count := 0
	for _,i := range nums{
		for _,value := range i{
			onenums[count] = value
			count++
		}
	}


	newnums := make([][]int, r)
	for i := 0; i < r; i++ {
		newnums[i] = make([]int, c)
		copy(newnums[i], onenums[c*i:c*(i+1)])
	}
	return newnums

}

/*
这里还是强调一下二维数组的初始化、
ans := make([][]int, r)
    for i:=0; i<r; i++ {
        ans[i] = make([]int, c)
    }
*/

File: 566. Reshape the Matrix/566.py
class Solution:
    def matrixReshape(self, nums, r: int, c: int) :
        import  copy
        if len(nums)*len(nums[0]) != r*c:
            return copy.deepcopy(nums)
        tmplist = []
        for i in nums:
            tmplist += i
        finlist = []
        for i in range(r):
            finlist.append(tmplist[c*i:c*(i+1)])
        return finlist

    # 这个方法真的是很好，很python，写法可读性虽然不好，但我喜欢
    def matrixReshape1(self, nums, r, c):
        return nums if len(sum(nums, [])) != r * c else map(list, zip(*([iter(sum(nums, []))]*c)))

    # 这个方法把列表表达式用到了极致吧...这个双循环列表表达式还是第一次见
    def matrixReshape2(self, nums, r, c):
        if r * c != len(nums) * len(nums[0]):
                return nums
        else:
            items = [y for x in nums for y in x]
            return [items[x*c : ((x+1)*c)] for x in range(r)]

a = Solution()
print(a.matrixReshape([[1,2],[3,4]],2,4))
